File: 2022/day3.ts
import * as fs from "fs";
function part1(): number {
    let count = 0
    let bags = fs.readFileSync("in.txt", "utf8").split("\r\n")
    for (let i in bags) {
        let map = new Map()
        for (let j = 0; j < bags[i].length; j++) {
            if (j < Math.floor(bags[i].length / 2)) {
                map.set(bags[i][j], true)
            } else {
                if (map.has(bags[i][j])) {
                    count += bags[i][j].charCodeAt(0) - 64 < 27 ? bags[i][j].charCodeAt(0) - 38 : bags[i][j].charCodeAt(0) - 96
                    break
                }
            }
        }
    }
    return count
}

function part2(): number {
    let count = 0
    let bags = fs.readFileSync("in.txt", "utf8").split("\r\n")
    for (let i = 0; i < Math.floor(bags.length / 3); i++) {
        let map = new Map()
        for (let j = 0; j < 3; j++) {
            for (let k = 0; k < bags[i * 3 + j].length; k++) {
                if (j == 0) {
                    map.set(bags[i * 3 + j][k], 1)
                } else if (map.has(bags[i * 3 + j][k])) {
                    if (j == 2 && map.get(bags[i * 3 + j][k]) > 1) {
                        count += bags[i * 3 + j][k].charCodeAt(0) - 64 < 27 ? bags[i * 3 + j][k].charCodeAt(0) - 38 : bags[i * 3 + j][k].charCodeAt(0) - 96
                        break
                    } else if (j == 1) {
                        map.set(bags[i * 3 + j][k], 2)
                    }
                }
            }
        }
    }
    return count
}

console.log(part1())
console.log(part2())

File: 2025/day3.py
def part1():
    count = 0
    for i in open("in.txt").read().split("\n"):
        nums = [0] * 2
        for j in range(len(i)):
            if j != len(i) - 1 and int(i[j]) > nums[0]:
                nums[0] = int(i[j])
                nums[1] = 0
            elif int(i[j]) > nums[1]:
                nums[1] = int(i[j])
        count += nums[0] * 10 + nums[1]
    return count

def part2():
    count = 0
    for i in open("in.txt").read().split("\n"):
        nums = [0] * 12
        for j in range(len(i)):
            clear = False
            for k in range(len(nums)):
                if clear:
                    nums[k] = 0
                elif j + (len(nums) - 1 - k) < len(i) and int(i[j]) > nums[k]:
                    nums[k] = int(i[j])
                    clear = True
        for j in range(len(nums)):
            count += nums[j] * 10 ** (len(nums) - 1 - j)
    return count

if __name__ == "__main__":
    print("Part 1:", part1())
    print("Part 2:", part2())